src/truncated-negative-binomial-distribution.cpp

```diff
@@ -0,0 +1,505 @@
+                          double a, double b, bool& throw_warning) {
+#ifdef IEEE_754
+  if (ISNAN(x) || ISNAN(size) || ISNAN(prob) || ISNAN(a) || ISNAN(b))
+    return x+size+prob+a+b;
+#endif
+  if (size <= 0.0 || !VALID_PROB(prob) || b < a) {
+    throw_warning = true;
+    return NAN;
+  }
+  
+  if (x < 0.0 || x <= a)
+    return 0.0;
+  if (x > b || !R_FINITE(x))
+    return 1.0;
+  
+  double pa, pb;
+  pa = R::pnbinom(a, size, prob, true, false);
+  pb = R::pnbinom(b, size, prob, true, false);
+  
+  return (R::pnbinom(x, size, prob, true, false) - pa) / (pb-pa);
+}
+
+inline double cdf_tnbinom_mu(double x, double size, double mu, 
+                             double a, double b, bool& throw_warning) {
+#ifdef IEEE_754
+  if (ISNAN(x) || ISNAN(size) || ISNAN(mu) || ISNAN(a) || ISNAN(b))
+    return x+size+mu+a+b;
+#endif
+  if (size <= 0.0 || mu < 0.0 || b < a) {
+    throw_warning = true;
+    return NAN;
+  }
+  
+  if (x < 0.0 || x <= a)
+    return 0.0;
+  if (x > b || !R_FINITE(x))
+    return 1.0;
+  
+  double pa, pb;
+  pa = R::pnbinom_mu(a, size, mu, true, false);
+  pb = R::pnbinom_mu(b, size, mu, true, false);
+  
+  return (R::pnbinom_mu(x, size, mu, true, false) - pa) / (pb-pa);
+}
+
+inline double invcdf_tnbinom(double p, double size, double prob, 
+                             double a, double b, bool& throw_warning) {
+#ifdef IEEE_754
+  if (ISNAN(p) || ISNAN(size) || ISNAN(prob) || ISNAN(a) || ISNAN(b))
+    return p+size+prob+a+b;
+#endif
+  if (size <= 0.0 || !VALID_PROB(prob) || b < a || !VALID_PROB(p)) {
+    throw_warning = true;
+    return NAN;
+  }
+  
+  if (p == 0.0)
+    return std::max(a, 0.0);
+  if (p == 1.0)
+    return b;
+  
+  double pa, pb;
+  pa = R::pnbinom(a, size, prob, true, false);
+  pb = R::pnbinom(b, size, prob, true, false);
+  
+  return R::qnbinom(pa + p*(pb-pa), size, prob, true, false);
+}
+
+inline double invcdf_tnbinom_mu(double p, double size, double mu, 
+                                double a, double b, bool& throw_warning) {
+#ifdef IEEE_754
+  if (ISNAN(p) || ISNAN(size) || ISNAN(mu) || ISNAN(a) || ISNAN(b))
+    return p+size+mu+a+b;
+#endif
+  if (size <= 0.0 || mu < 0.0 || b < a || !VALID_PROB(p)) {
+    throw_warning = true;
+    return NAN;
+  }
+  
+  if (p == 0.0)
+    return std::max(a, 0.0);
+  if (p == 1.0)
+    return b;
+  
+  double pa, pb;
+  pa = R::pnbinom_mu(a, size, mu, true, false);
+  pb = R::pnbinom_mu(b, size, mu, true, false);
+  
+  return R::qnbinom_mu(pa + p*(pb-pa), size, mu, true, false);
+}
+
+inline double rng_tnbinom(double size, double prob, double a, double b,
+                          bool& throw_warning) {
+  if (ISNAN(size) || ISNAN(prob) || ISNAN(a) || ISNAN(b) ||
+      size <= 0.0 || !VALID_PROB(prob) || b < a) {
+    throw_warning = true;
+    return NA_REAL;
+  }
+  
+  double u, pa, pb;
+  pa = R::pnbinom(a, size, prob, true, false);
+  pb = R::pnbinom(b, size, prob, true, false);
+  
```

Comment:
This part repeats in all methods, so probably could be extracted as a separate method.

Reply:
@twolodzko Please elaborate. Are you suggesting something like below?

```cpp
double u, pa, pb;
tie(pa, pb) = pnbinom_ends(a, b, size, prob);
```

We'd need a second `pnbinom_ends_mu`. Maybe you can propose a better name.

Reply:
I mean that calling `R::pnbinom_mu` and `R::pnbinom` with the same parameters happens several times in the code so they can be extracted as separate functions.